tex/gregoriotex-signs.tex

```diff
@@ -2695,16 +2699,113 @@
+  \global\gre@alteration@effect@linefalse
+  \IfStrEq{#1}{}{\gre@error{Empty argument for \protect\gresetalterationeffect. Allowed arguments are: 'note', 'line'}}{%
+    \def\gre@alteration@effect@process##1{%
+      \IfStrEqCase{##1}{%
+        {note}{\global\gre@alteration@effect@notetrue}%
+        {line}{\global\gre@alteration@effect@linetrue}%
+      }[\gre@error{Unrecognized argument '##1' for \protect\gresetalterationeffect. Allowed arguments are: 'note', 'line'}]}%
+    \gre@parsecommas{#1}{\gre@alteration@effect@process}%
+  }%
+}
+
+% Test whether an alteration should be visible.
+% #1: 0 for current alteration, 1 for next alteration.
+% #2: alteration type
+% Sets \ifgre@alteration@isvisibletrue or \ifgre@alteration@isvisiblefalse.
+\newif\ifgre@alteration@isvisible
+\def\gre@alteration@isvisible#1#2{%
+  \global\gre@alteration@isvisiblefalse
+  \ifnum#2>6\relax % soft
+    % alterationeffect includes "line"
+    \ifgre@alteration@effect@line
+      \gre@debugmsg{alteration}{line}%
+      \ifcase\directlua{gregoriotex.is_first_alteration(#1)}\relax
+        % 0: don't know
+        \global\gre@alteration@isvisibletrue
+      \or
+        % 1: first in line
+        \ifnum#2=8\relax % natural: don't print
+        \else
+          \global\gre@alteration@isvisibletrue
+        \fi
+      \or
+        % 2: different from previous
+        \global\gre@alteration@isvisibletrue
+      \or
+        % 3: not first
+      \fi
+    \fi
+    % alterationeffect includes "note"
+    \ifgre@alteration@effect@note
+      \gre@debugmsg{alteration}{note}%
+      \ifnum#2=8\relax % natural: don't print
+      \else
+        \global\gre@alteration@isvisibletrue
+      \fi
+    \fi
+  \else % hard: always visible
+    \global\gre@alteration@isvisibletrue
+  \fi
+}
+
+\def\gre@alteration@visible#1#2#3#4#5#6{%
+  \gre@trace{gre@alteration@visible{#1}{#2}{#3}{#4}{#5}{#6}}%
+  \ifcase#2\relax
+    \or\def\gre@alteration@char{\gre@fontchar@flat}\def\gre@alteration@hole{\gre@fontchar@flathole}% flat
+    \or\def\gre@alteration@char{\gre@fontchar@natural}\def\gre@alteration@hole{\gre@fontchar@naturalhole}% natural
+    \or\def\gre@alteration@char{\gre@fontchar@sharp}\def\gre@alteration@hole{\gre@fontchar@sharphole}% sharp
+    \or\def\gre@alteration@char{\gre@fontchar@flatparen}\def\gre@alteration@hole{\gre@fontchar@flatparenhole}% parenthesized flat
+    \or\def\gre@alteration@char{\gre@fontchar@naturalparen}\def\gre@alteration@hole{\gre@fontchar@naturalparenhole}% parenthesized natural
+    \or\def\gre@alteration@char{\gre@fontchar@sharpparen}\def\gre@alteration@hole{\gre@fontchar@sharpparenhole}% parenthesized sharp
+    \else\gre@error{Invalid alteration type}% soft alterations should have been changed to hard
+  \fi
+  \setbox\gre@box@temp@width=\hbox
+    attr \gre@attrid@alteration@pitch=#1\relax
+    attr \gre@attrid@alteration@type=#2\relax
+    {\gre@pointandclick{\gre@alteration@char}{#6}}%
+  \ifnum#3=0\relax %
```

Comment:
Coding style question: `%` is probably not needed after `\relax` Do we want to be consistent with omitting the `%`? 

`:2764, :2795, :2805, :2806` vs `:2808`

Also the `\relax % Comment` construction shows the gobbled whitespace VS. `\relax% Comment` For readability I am fine with `\relax % Comment` and omitting the `%` after other `\relax`'s that have no comment.

Reply:
This happens a lot in existing code (>1000 times) and I was following the convention. I agree that it does not need to be there, but its removal should perhaps be another (huge) PR.

Reply:
I agree that this sort of style thing should await a different PR.